flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/OuterJoinRecordStateViews.java

```diff
@@ -160,6 +161,17 @@ static void processElements(JoinRecordStateView otherView, Collector<RowData> co
                 collect.collect(outRow);
             }
         }
+
+        if (!isAntiJoin && StreamingPerfOptimizationUtil.isEligibleForNullEquijoinOptimization(
+                StreamingPerfOptimizationUtil.isKeyAnyNulls(thisRow),
```

Comment:
thisRow is the complete row here, not just the key -- you'll be incorrectly capturing non-key payload columns which are null but the key itself is not null, which is incorrect
the right place to check for "are there nulls in the key for this row" is in the loop that iterates keys in the backend right next to the call that sets backend key context, you alternatively you pass the join key into there

Reply:
Ah good catch! Thank you - fixed!